adapter/src/components/LoginModal.js

```diff
@@ -6,57 +6,133 @@ import {
+    const [error, setError] = useState(null)
+    const [isLoggingIn, setIsLoggingIn] = useState(false)
 
     const isValid = (val) => val && val.length >= 2
+    const getSignInButtonText = ({ loginApp, isLoggingIn }) => {
+        if (!loginApp) {
+            return isLoggingIn ? i18n.t('Signing in...') : i18n.t('Sign in')
+        }
+        return isLoggingIn ? i18n.t('Going to app...') : i18n.t('Go to app')
+    }
 
     const onSubmit = async (e) => {
         e.preventDefault()
         setIsDirty(true)
-        if (isValid(server) && isValid(username) && isValid(password)) {
+        if (
+            isValid(server) &&
+            ((isValid(username) && isValid(password)) || loginApp)
+        ) {
+            setIsLoggingIn(true)
             if (!staticUrl) {
                 // keep the localStorage value here -- it's still used in some
                 // obscure cases, like in the cypress network shim
                 window.localStorage.DHIS2_BASE_URL = server
                 await setBaseUrlByAppName({ appName, baseUrl: server })
+                if (loginApp) {
+                    window.location.reload()
+                }
             }
-            try {
-                await post(
-                    `${server}/dhis-web-commons-security/login.action`,
-                    `j_username=${encodeURIComponent(
-                        username
-                    )}&j_password=${encodeURIComponent(password)}`
-                )
-            } catch (e) {
-                console.log(
-                    'TODO: This will always error and cancel the request until we get a real login endpoint!'
-                )
-            }
 
-            // TODO: Hacky solution... this shouldn't require a reload
-            window.location.reload()
+            const useNewLoginAPI = await getUseNewLoginAPI(server)
+
+            if (useNewLoginAPI) {
+                loginWithNewEndpoints({
+                    server,
+                    username,
+                    password,
+                    setError,
+                    setIsLoggingIn,
+                })
+            } else {
+                loginWithOldEndpoints({ server, username, password })
+            }
         }
     }
 
     return (
         <Modal open small dataTest="dhis2-adapter-loginmodal">
+            <style jsx>{styles}</style>
             <form onSubmit={onSubmit}>
-                <ModalTitle>{i18n.t('Please sign in')}</ModalTitle>
+                <ModalTitle>
+                    {!loginApp
+                        ? i18n.t('Please sign in')
+                        : i18n.t('Specify server')}
+                </ModalTitle>
 
                 <ModalContent>
+                    {error && (
+                        <div className="errorNotification">
+                            <NoticeBox error title={i18n.t('Could not log in')}>
+                                {error?.message || error?.details?.message}
```

Comment:
we actually seem to not be throwing the error so that it can be properly caught here...so I could update that or we could just skip displaying a message for now

Reply:
I think it would be nice to show a message! I played around a little to see what would be needed, and I think it just needs this at line 17 in `api.js`:
```diff
-                    reject('Request failed ' + response.statusText)
+                    response.json().then(json => {
+                        reject(json)
+                    })
```

Reply:
lol the message from the backend feels slightly harsh

<img width="417" alt="Screenshot 2024-04-26 at 11 36 00 AM" src="https://github.com/dhis2/app-platform/assets/49666798/df75a91d-a00b-4b69-97f8-3bd4f6024ab0">

Reply:
@KaiVandivier:  how about 
```
                if (response.status !== 200) {
                    response
                        .json()
                        .then((json) => {
                            reject(json)
                        })
                        .catch(() => {
                            reject('Request failed ' + response.statusText)
                        })
                    return
                }
```

I think there could be api requests returning non-json responses, but the json() method should throw an error and then we can catch and reject with `response.statusText` like before?

Reply:
Good call! That sounds perfect 👌